FIRST ITERATION OF ENUMERABLES READY!

angel-mora opened the pull request:
# Project not completed, requesting revision for some hint

All methods working but multiply-els. Requesting some guidance to address cases different than the single case addressed in tests.

angel-mora committed:
added basic rb file, need to add rubocop.yml

angel-mora committed:
added initial files, need to verify rubocop working properly

angel-mora committed:
playing around with each method

angel-mora committed:
added first method with rubocop observations addressed. However, Prefer each over for comment not addressed as replicating each method

angel-mora committed:
working on simple version of methods, added my_select

angel-mora committed:
added a couple more methods

angel-mora committed:
added first iteration of enumerables without considering all cases

angel-mora committed:
debugging

angel-mora committed:
debugging

mikenath223 reviewed (changes requested):
### Hi @angel-mora 👋

Great job on getting to this stage🚀

Let's do the following:

- [ ] Kindly fix the linting issues on your repo. 

 - [ ] Avoid using a return statement in the last line of each method as ruby automatically returns the last line.

- [ ] The following methods are expected to return an enum when no block is passed: `my_each`, `my_each_with_index`, `my_select`, `my_any?`, `my_map`. Kindly make the adjustment.

- [ ] The following methods have not been structured to take a Regex or a ruby Class and return a result: `my-all?`, `my_any?` and  `my_none?`.

- [ ]  In addition kindly avoiding using a puts statement in your methods.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

angel-mora committed:
separated tests and enumerables, fixed most styling linters, about to add not considered cases

angel-mora committed:
need to add more cases for methods, fix rubocop linter issues

angel-mora committed:
added my_multiply_els method. Next steps are 1.fix rubocop 2.add and test cases for methods

angel-mora committed:
solved my_select to return array instead of list of elements

angel-mora committed:
added my each and my select methods considering remaining cases

angel-mora committed:
added my_all method

angel-mora committed:
added my_count, debugged a bit

angel-mora committed:
added my_map method

angel-mora committed:
debbuging, all methods finished

angel-mora committed:
project finished

angel-mora committed:
rubocop issue

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+      yield(aux[x])
+      x += 1
+    end
+  end
```
`my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods)) ``
- each can perform with hash. But your one is not working with it.
- After calling the block, the method returns self. But your one is not doing it.
```ruby
[1, 2, 3].my_each {|num| num}  #should return [1, 2, 3]
(1...4).my_each {|num| num}  #should return (1...4)
{a: 4, b: 3}.my_each {|item| item}  #should return {a: 4, b: 3}
```
So I was trying to refactor your code and came up with the following solution.
```ruby
def my_each
    return to_enum unless block_given?

    aux = is_a?(Array) ? self : to_a
    x = 0
    while x < aux.length
      yield(aux[x])
      x += 1
    end
    self
end
```
So if you think that it is covering your issue, you can add it instead of your my_each.

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+      yield(self[i], i)
+      i += 1
+    end
+  end
```
`my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- each_with_index can perform with hash. But your one is not working with it.
- After calling the block, the method returns self. But your one is not doing it.
```ruby
[1, 2, 3].my_each_with_index {|num, idx| num}  #should return [1, 2, 3]
(1...4).my_each_with_index {|num, idx| num}  #should return (1...4)
{a: 4, b: 3}.my_each_with_index {|item, idx| item}  #should return {a: 4, b: 3}
```
Well to handle this issue you can take help from the previous comment made in my_each block. The solution is pretty much the same with an extra block variable.

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+      i += 1
+    end
+    true
+  end
```
`my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods)) 
- my_all is also function with range. Should interact as the all?
```ruby
puts (1...3).my_all? # getting error but should return true
puts (1...3).all? # returns true.
```
:rocket: Hint -> The possible solution is added on the sample code that I have given above :point_up:

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+      i += 1
+    end
+    false
+  end
```
`my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- my_any can interact with range too.
```ruby
puts (1...3).my_any? # getting error but should return true
puts (1...3).any? # returns true.
```
Same hint as above :point_up:

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+      i += 1
+    end
+    true
+  end
```
`my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- can interact with range too. so the hint is pointed above :point_up:

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+      counter = length
+    end
+    counter
+  end
```
`my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- It can interact with range too, Well the hint is attached above.
```ruby
(1...3).my_count # should return 2

```

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+
+  def my_multiply_els
+    my_inject { |x, y| x * y }
+  end
```
it should be named as multiply_els. And should be placed outside of the module. Cause it is not an enumerable method.

shshamim63 commented on enumerables.rb:
```diff
@@ -0,0 +1,143 @@
+    modified
+  end
+
+  def my_inject(*args)
+    if args[0].is_a? Integer
+      accumulator = args[0]
+      symbol = args[1]
+    else
+      symbol = args[0]
+    end
+
+    if block_given? && args.size == 1
+      my_each { |x| accumulator = yield(accumulator, x) }
+    elsif args.size == 2
+      my_each { |x| accumulator = accumulator.send(symbol, x) }
+    elsif block_given?
+      my_each { |x| accumulator = accumulator ? yield(accumulator, x) : x }
+    else
+      my_each { |x| accumulator = accumulator ? accumulator.send(symbol, x) : x }
+    end
+    accumulator
+  end
```
`my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- raises a "LocalJumpError" when no block or argument is given

angel-mora replied:
I'm afraid regular inject method also raises "LocalJumpError".
I tried to did it without block nor argument and it didn't work.
```
[100, 200, 1000].inject
```
returns
```
Traceback (most recent call last):
    2: from main.rb:1:in `<main>'
    1: from main.rb:1:in `inject'
main.rb:1:in `each': no block given (LocalJumpError)
```

shshamim63 reviewed (changes requested):
## Status :dart: Required Changes :dart: 
___
Hi @angel-mora :wave: 
Wow, you have covered almost all the changes :clap: .Just a little bit more to go. And I am sure that you will complete them easily. Let's cover some edge cases.
___
### :spades: Mandatory Changes :spades: 
___
- The following methods having some corner cases. Please read the instructions added to theirs respective block comments and resolve them.
  - my_each
  - my_each_with_index
  - my_all?
  - my_any?
  - my_none?
  - my_count
  - my_inject
  - multiply_els
___
### :clubs: Optional Changes :clubs: 
___
- Your readme is not well descriptive. And the previous reviewer missed it to mention it. The readme can contain what are the methods you have cloned here. How to use them, how to contact you, do you accept PR, and so on. Remember a user always starts from a readme.
___
:warning: You did not follow the correct git flow. Check the remember_to section of this [link](
https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421589/). The PR should have been marked as invalid but as the previous reviewer started to review on it, I am not marking it as invalid. Please make sure that you follow the correct flow in your next project.
___
##### Please resolve the issues mentioned above and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

angel-mora committed:
debugging

angel-mora committed:
changes addressed, ready to new review

angel-mora commented:
# UPDATED PULL REQUEST

## Changes addressed

- Added an evaluator for object input in my_all, my_none, my_any and my_count methods.
- Inject **does** return an error without block or argument not given.
- Fixed multiply_els method.
- Added extra tests in _tests.rb_ file.
- Readme file modified.
- Acknowledged issue regarding Git Flow. Should have created a development branch right after master and then a branch feature from development. 

Thank you for your helpful feedback.

angel-mora committed:
added readme file

serdg0 reviewed (changes requested):
### Hi @angel-mora! :peace_symbol: 

**Congratulations!** You have done a great job!
Although we need to address some issues before approving it:

- [ ] my_each_with_index: Consider when a **range** and a **hash** is passed.


- [ ] my_all: Consider the case where you pass a pattern other than Regex. We tested it with an array where all of it's elements are the same number, and it returned false.


- [ ] my_any: When no block or argument is given returns false if at least one of the collection is not true. Also, consider the case where you pass a pattern other than Regex. We tested it with a string and compared it with the original method, and the result was not the same.


- [ ] my_none: Consider the case where you pass a pattern other than Regex. It should return false only if one of the collection matches the pattern.


- [ ] my_inject should raise a LocalJumpError when no block or argument is given. [Here](https://airbrake.io/blog/ruby-exception-handling/localjumperror) is a link that I hope it helps you handling this issue.


- [ ] multiply_els can accept an array as an argument. For example: Calling it like this: multiply_els [2, 4, 5] should return 40.

 Address those issues and submit for another CRR.

Here is a little **hint** that might help you understand better those issues: Compare the output of your methods with the original one. That can be made in the terminal :smiley: 

Keep the good work and...

### **Happy Coding!** :computer: :nerd_face:
**_Sergio Diaz_**

angel-mora committed:
fixed issues by TSE

rOluochKe reviewed (approved):
Hello Team!
### You have done well and for making changes from the previous reviews :clap: Your project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one.
- Happy coding!

angel-mora committed:
add rspec init and enumerables_tests.rb

CalyCherkaoui committed:
adding require_relative

CalyCherkaoui committed:
updating tests

CalyCherkaoui committed:
adding my_each method test

CalyCherkaoui committed:
fixing spec

CalyCherkaoui committed:
testing the my_each method

angel-mora committed:
added my_each_with_index tests

CalyCherkaoui committed:
testing my_select method

angel-mora committed:
my_all tests started, completed first two cases

CalyCherkaoui committed:
Testing my_all? method

CalyCherkaoui committed:
Testing my any method

angel-mora committed:
added my_none test

angel-mora committed:
added my_count

CalyCherkaoui committed:
testing My_map method

angel-mora committed:
added my_inject tests, need to review 2 failures

angel-mora commented:
# Rspec Tests for Enumerables

Submitting Enumerables tests. Still have some questions available on line comments.

nikasulo reviewed (changes requested):
# Nice work team!
You've done well building this, however, you haven't setup your `gemfile` properly, and as such, I wasn't able to install any gems. Please setup properly, run `bundle install` and be sure it runs smoothly before you request another review. I wish you the very best and see you at the top 🚀
